kombu/connection.py

```diff
@@ -1035,6 +1035,29 @@ def is_evented(self):
 BrokerConnection = Connection
 
 
+class PooledConnection:
+    """Wraps :class:`kombu.Connection`.
+
+    This wrapper modifies :meth:`kombu.Connection.__exit__` to close the connection
+    in case any exception occurred while the context was active.
+    """
+
+    def __init__(self, connection, pool):
+        self._connection = connection
+        self._pool = pool
+
+    def __getattr__(self, attr):
+        return getattr(self._connection, attr)
+
+    def __enter__(self):
+        return self._connection.__enter__()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_type is not None and self._pool.limit:
+            self._pool.replace(self._connection)
```

Comment:
The lack of `return` is intentional?

Reply:
Updated the class to now be a child class of `Connection`. 
The forwarding using `getattr` was not particularly nice anyway and while it passed the unit tests, it caused issues with celery.